trie/binary.go

```diff
@@ -0,0 +1,483 @@
+
+	// 0_32
+	zero32 = []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
+)
+
+func newBinKey(key []byte) binkey {
+	bits := make([]byte, 8*len(key))
+	for i := range bits {
+
+		if key[i/8]&(1<<(7-i%8)) == 0 {
+			bits[i] = 0
+		} else {
+			bits[i] = 1
+		}
+	}
+
+	return binkey(bits)
+}
+func min(i, j int) int {
+	if i < j {
+		return i
+	}
+	return j
+}
+func (b binkey) commonLength(other binkey) int {
+	for i := 0; i < len(b) && i < len(other); i++ {
+		if b[i] != other[i] {
+			return i
+		}
+	}
+	return min(len(b), len(other))
+}
+func (b binkey) samePrefix(other binkey, off int) bool {
+	return bytes.Equal(b[off:off+len(other)], other[:])
+}
+
+func (s store) Len() int { return len(s) }
+func (s store) Less(i, j int) bool {
+	for b := 0; b < len(s[i].key) && b < len(s[j].key); b++ {
+		if s[i].key[b] != s[j].key[b] {
+			// if s[j].key.Bit(b) is true, then it is
+			// the greater value of the two.
+			return s[j].key[b] == 1
+		}
+	}
+
+	// Keys are equal on their common length, the shortest
+	// is the smaller one.
+	return len(s[i].key) < len(s[j].key)
+}
+func (s store) Swap(i, j int) {
+	temp := s[i]
+	s[i] = s[j]
+	s[j] = temp
+}
+
+func NewBinTrie() BinaryNode {
+	return empty(struct{}{})
+}
+
+func (bt *BinaryTrie) TryGet(key []byte) ([]byte, error) {
+	bk := newBinKey(key)
+	off := 0
+
+	var currentNode *branch
+	switch bt.root.(type) {
+	case empty:
+		return nil, errKeyNotPresent
+	case *branch:
+		currentNode = bt.root.(*branch)
+	case hashBinaryNode:
+		return nil, errReadFromHash
+	}
+
+	for {
+		if !bk.samePrefix(currentNode.prefix, off) {
+			return nil, errKeyNotPresent
+		}
+
+		// If it is a leaf node, then the a leaf node
+		// has been reached, and the value can be returned
+		// right away.
+		if currentNode.value != nil {
+			return currentNode.value, nil
+		}
+
+		// This node is a fork, get the child node
+		var childNode *branch
+		if bk[off+len(currentNode.prefix)] == 0 {
+			childNode = bt.resolveNode(currentNode.left, bk, off+1, nil)
+		} else {
+			childNode = bt.resolveNode(currentNode.right, bk, off+1, nil)
+		}
+
+		// if no child node could be found, the key
+		// isn't present in the trie.
+		if childNode == nil {
+			return nil, errKeyNotPresent
+		}
+		off += len(currentNode.prefix) + 1
+		currentNode = childNode
+	}
+}
+
+// Hash calculates the hash of an expanded (i.e. not already
+// hashed) node.
+func (br *branch) Hash() []byte {
+	return br.hash(0)
+}
+
+func (br *branch) hash(off int) []byte {
+	hasher := newHasher(false)
+	defer returnHasherToPool(hasher)
+	hasher.sha.Reset()
+
+	hash := make([]byte, 32)
+	if br.value == nil {
+		// This is a branch node, so the rule is
+		// branch_hash = hash(left_root_hash || right_root_hash)
+		hasher.sha.Write(br.left.hash(off + len(br.prefix) + 1))
+		hasher.sha.Write(br.right.hash(off + len(br.prefix) + 1))
+		hasher.sha.Read(hash)
+		hasher.sha.Reset()
+	} else {
```

Comment:
You don't need to borrown the `hasher` until later. As is, the top level will 'check out' a hasher, then go into children, and all of them will check out a hasher from the pool, and then start returning the hashes. 

Instead, you can do 

```suggestion
	if br.value == nil {
		// This is a branch node, so the rule is
		// branch_hash = hash(left_root_hash || right_root_hash)
		left := br.left.hash(off + len(br.prefix) + 1)
		right := br.right.hash(off + len(br.prefix) + 1)
		hasher := newHasher(false)
		hasher.sha.Reset()
		hasher.sha.Write(left)
		hasher.sha.Write(right)
		hasher.sha.Read(hash)
		returnHasherToPool(hasher)
	} else {
```